File: src/runtime/virtcontainers/physical_endpoint.go
//go:build linux

// Copyright (c) 2018 Intel Corporation
//
// SPDX-License-Identifier: Apache-2.0
package virtcontainers

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/containernetworking/plugins/pkg/ns"
	"github.com/kata-containers/kata-containers/src/runtime/pkg/device/config"
	"github.com/kata-containers/kata-containers/src/runtime/pkg/device/drivers"
	resCtrl "github.com/kata-containers/kata-containers/src/runtime/pkg/resourcecontrol"
	persistapi "github.com/kata-containers/kata-containers/src/runtime/virtcontainers/persist/api"
	vcTypes "github.com/kata-containers/kata-containers/src/runtime/virtcontainers/types"
	"github.com/safchain/ethtool"
	logrus "github.com/sirupsen/logrus"
	"github.com/vishvananda/netlink"
)

var physicalTrace = getNetworkTrace(PhysicalEndpointType)

// PhysicalEndpoint gathers a physical network interface and its properties
type PhysicalEndpoint struct {
	IfaceName          string
	IsVFIO             bool
	HardAddr           string
	EndpointProperties NetworkInfo
	EndpointType       EndpointType
	BDF                string
	Driver             string
	VendorDeviceID     string
	PCIPath            vcTypes.PciPath
	CCWDevice          *vcTypes.CcwDevice
	NetPair            NetworkInterfacePair
	BusType            string
	RxRateLimiter      bool
	TxRateLimiter      bool
}

// Properties returns the properties of the physical interface.
func (endpoint *PhysicalEndpoint) Properties() NetworkInfo {
	return endpoint.EndpointProperties
}

// HardwareAddr returns the mac address of the physical network interface.
func (endpoint *PhysicalEndpoint) HardwareAddr() string {
	return endpoint.HardAddr
}

// Name returns name of the physical interface.
func (endpoint *PhysicalEndpoint) Name() string {
	return endpoint.IfaceName
}

// Type indentifies the endpoint as a physical endpoint.
func (endpoint *PhysicalEndpoint) Type() EndpointType {
	return endpoint.EndpointType
}

// PciPath returns the PCI path of the endpoint.
func (endpoint *PhysicalEndpoint) PciPath() vcTypes.PciPath {
	return endpoint.PCIPath
}

// SetPciPath sets the PCI path of the endpoint.
func (endpoint *PhysicalEndpoint) SetPciPath(pciPath vcTypes.PciPath) {
	endpoint.PCIPath = pciPath
}

// CcwDevice returns the CCW device of the endpoint.
func (endpoint *PhysicalEndpoint) CcwDevice() *vcTypes.CcwDevice {
	return endpoint.CCWDevice
}

// SetCcwDevice sets the CCW device of the endpoint.
func (endpoint *PhysicalEndpoint) SetCcwDevice(ccwDev vcTypes.CcwDevice) {
	endpoint.CCWDevice = &ccwDev
}

// SetProperties sets the properties of the physical endpoint.
func (endpoint *PhysicalEndpoint) SetProperties(properties NetworkInfo) {
	endpoint.EndpointProperties = properties
}

// NetworkPair returns the network pair of the endpoint.
func (endpoint *PhysicalEndpoint) NetworkPair() *NetworkInterfacePair {
	return &endpoint.NetPair
}

// Attach for physical endpoint binds the physical network interface to
// vfio-pci and adds device to the hypervisor with vfio-passthrough.
func (endpoint *PhysicalEndpoint) Attach(ctx context.Context, s *Sandbox) error {
	span, ctx := physicalTrace(ctx, "Attach", endpoint)
	defer span.End()
	if endpoint.IsVFIO {
		// Unbind physical interface from host driver and bind to vfio
		// so that it can be passed to qemu.
		vfioPath, err := bindNICToVFIO(endpoint)
		if err != nil {
			return err
		}
		c, err := resCtrl.DeviceToCgroupDeviceRule(vfioPath)
		if err != nil {
			return err
		}
		d := config.DeviceInfo{
			ContainerPath: vfioPath,
			DevType:       string(c.Type),
			Major:         c.Major,
			Minor:         c.Minor,
			ColdPlug:      true,
			Port:          s.config.HypervisorConfig.ColdPlugVFIO,
		}
		_, err = s.AddDevice(ctx, d)
		return err
	} else {
		h := s.hypervisor
		if err := xConnectVMNetwork(ctx, endpoint, h); err != nil {
			networkLogger().WithError(err).Error("Error bridging physical endpoint")
			return err
		}
		if err := h.AddDevice(ctx, endpoint, NetDev); err != nil {
			networkLogger().WithError(err).Error("Error adding physical endpoint device")
			return err
		}
		return nil
	}
}

// Detach for physical endpoint unbinds the physical network interface from vfio-pci
// and binds it back to the saved host driver.
func (endpoint *PhysicalEndpoint) Detach(ctx context.Context, netNsCreated bool, netNsPath string) error {
	span, _ := physicalTrace(ctx, "Detach", endpoint)
	defer span.End()
	if endpoint.IsVFIO {
		// Bind back the physical network interface to host.
		// We need to do this even if a new network namespace has not
		// been created by virtcontainers.
		// We do not need to enter the network namespace to bind back the
		// physical interface to host driver.
		return bindNICToHost(endpoint)
	} else {
		// The network namespace would have been deleted at this point
		// if it has not been created by virtcontainers.
		if !netNsCreated {
			return nil
		}
		return doNetNS(netNsPath, func(_ ns.NetNS) error {
			return xDisconnectVMNetwork(ctx, endpoint)
		})
	}
}

// HotAttach for physical endpoint not supported yet
func (endpoint *PhysicalEndpoint) HotAttach(ctx context.Context, s *Sandbox) error {
	span, ctx := physicalTrace(ctx, "HotAttach", endpoint)
	defer span.End()
	if endpoint.IsVFIO {
		// Unbind physical interface from host driver and bind to vfio
		// so that it can be passed to the hypervisor.
		vfioPath, err := bindNICToVFIO(endpoint)
		if err != nil {
			return err
		}
		c, err := resCtrl.DeviceToCgroupDeviceRule(vfioPath)
		if err != nil {
			return err
		}
		d := config.DeviceInfo{
			ContainerPath: vfioPath,
			DevType:       string(c.Type),
			Major:         c.Major,
			Minor:         c.Minor,
			ColdPlug:      false,
		}
		_, err = s.AddDevice(ctx, d)
		return err
	} else {
		h := s.hypervisor
		if err := xConnectVMNetwork(ctx, endpoint, h); err != nil {
			networkLogger().WithError(err).Error("Error bridging physical endpoint (hotplug)")
			return err
		}
		if _, err := h.HotplugAddDevice(ctx, endpoint, NetDev); err != nil {
			networkLogger().WithError(err).Error("Error hotplugging physical endpoint device")
			return err
		}
		return nil
	}
}

// HotDetach for physical endpoint not supported yet
func (endpoint *PhysicalEndpoint) HotDetach(ctx context.Context, s *Sandbox, netNsCreated bool, netNsPath string) error {
	span, _ := physicalTrace(ctx, "HotDetach", endpoint)
	defer span.End()
	var vfioPath string
	var err error
	if endpoint.IsVFIO {
		if vfioPath, err = drivers.GetVFIODevPath(endpoint.BDF); err != nil {
			return err
		}
		c, err := resCtrl.DeviceToCgroupDeviceRule(vfioPath)
		if err != nil {
			return err
		}
		d := config.DeviceInfo{
			ContainerPath: vfioPath,
			DevType:       string(c.Type),
			Major:         c.Major,
			Minor:         c.Minor,
			ColdPlug:      false,
		}
		device := s.devManager.FindDevice(&d)
		s.devManager.RemoveDevice(device.DeviceID())
		// We do not need to enter the network namespace to bind back the
		// physical interface to host driver.
		return bindNICToHost(endpoint)
	} else {
		if !netNsCreated {
			return nil
		}
		span, ctx := vethTrace(ctx, "HotDetach", endpoint)
		defer span.End()
		if err := doNetNS(netNsPath, func(_ ns.NetNS) error {
			return xDisconnectVMNetwork(ctx, endpoint)
		}); err != nil {
			networkLogger().WithError(err).Warn("Error un-bridging virtual ep")
		}
		h := s.hypervisor
		if _, err := h.HotplugRemoveDevice(ctx, endpoint, NetDev); err != nil {
			return err
		}
		return nil
	}
}

// isPhysicalIface checks if an interface is a physical device by inspecting
// the link's ParentDevBus attribute. Returns true when the bus is "pci" or
// "vmbus". ParentDevBus is populated by the kernel via netlink
// and does not require sysfs access inside the network namespace.
func isPhysicalIface(link netlink.Link) bool {
	isParent := (link.Attrs().ParentDevBus == "pci" || link.Attrs().ParentDevBus == "vmbus")
	return isParent
}

var sysBusPath = "/sys/bus/"

// findSubordinateVF looks for a PCI virtual function (VF) paired with a
// VMBus netvsc NIC. On Azure with Accelerated Networking, the hv_netvsc
// driver bonds a Mellanox (or other) SR-IOV VF as a lower device for
// data-plane acceleration.
//
// The kata shim enters the pod's network namespace via setns(CLONE_NEWNET)
// but does NOT remount sysfs, so /sys/class/net/ still reflects the host's
// view.  Netlink and ethtool operate on the current network namespace
// regardless of mount namespace, so we use those instead for discovery.
//
// Returns the VF's PCI BDF (e.g. "8cd5:00:02.0") when found, or an
// empty string when no subordinate VF exists.
func findSubordinateVF(ifaceName string) (string, error) {
	links, err := netlink.LinkList()
	if err != nil {
		return "", fmt.Errorf("listing links to find subordinate VF: %w", err)
	}

	networkLogger().WithFields(logrus.Fields{
		"iface":      ifaceName,
		"link-count": len(links),
	}).Info("findSubordinateVF: enumerating netns links for PCI VF")

	ethHandle, err := ethtool.NewEthtool()
	if err != nil {
		return "", fmt.Errorf("creating ethtool handle for VF discovery: %w", err)
	}
	defer ethHandle.Close()

	for _, link := range links {
		attrs := link.Attrs()
		if attrs.Name == ifaceName {
			continue
		}
		if attrs.ParentDevBus != "pci" {
			networkLogger().WithFields(logrus.Fields{
				"link":           attrs.Name,
				"parent-dev-bus": attrs.ParentDevBus,
			}).Debug("findSubordinateVF: skipping non-PCI link")
			continue
		}

		bdf, err := ethHandle.BusInfo(attrs.Name)
		if err != nil {
			networkLogger().WithError(err).WithField("link", attrs.Name).Debug("findSubordinateVF: skipping PCI link, ethtool BusInfo failed")
			continue
		}

		networkLogger().WithFields(logrus.Fields{
			"parent-iface":       ifaceName,
			"subordinate-iface":  attrs.Name,
			"subordinate-hwaddr": attrs.HardwareAddr,
			"vf-bdf":             bdf,
		}).Info("Discovered subordinate PCI VF via netlink")

		return bdf, nil
	}

	networkLogger().WithField("iface", ifaceName).Info("findSubordinateVF: no subordinate PCI VF found in netns")
	return "", nil
}

// Get vendor and device id from pci space (sys/bus/pci/devices, or sys/bus/vmbus/devices, ...)
func getDevicesPath(link netlink.Link) string {
	return filepath.Join(sysBusPath, link.Attrs().ParentDevBus, "devices")
}

// Get vendor and device id from pci space (sys/bus/pci/devices/$BusDeviceInfo)
func getIfaceDevicePath(link netlink.Link, deviceInterfaceName string) (string, string, error) {
	if link.Attrs().ParentDevBus == "pci" {
		// Get ethtool handle to derive driver and bus
		ethHandle, err := ethtool.NewEthtool()
		if err != nil {
			return "", "", err
		}
		defer ethHandle.Close()
		// Get Bus info
		bdf, err := ethHandle.BusInfo(deviceInterfaceName)
		if err != nil {
			return "", "", err
		}
		// Get device by following symlink /sys/bus/pci/devices/$bdf
		return filepath.Join(getDevicesPath(link), bdf), bdf, nil
	} else if link.Attrs().ParentDevBus == "vmbus" {
		return filepath.Join(getDevicesPath(link), link.Attrs().ParentDev), link.Attrs().ParentDev, nil
	} else {
		return "", "", fmt.Errorf("unsupported ParentDevBus: %s", link.Attrs().ParentDevBus)
	}
}
func createPhysicalEndpoint(idx int, netInfo NetworkInfo, isFVIODisabled bool, interworkingModel NetInterworkingModel) (*PhysicalEndpoint, error) {
	attrs := netInfo.Link.Attrs()
	networkLogger().WithFields(logrus.Fields{
		"iface-name":     netInfo.Iface.Name,
		"iface-index":    netInfo.Iface.Index,
		"iface-mtu":      netInfo.Iface.MTU,
		"iface-hwaddr":   netInfo.Iface.HardwareAddr,
		"iface-flags":    netInfo.Iface.Flags,
		"link-type":      netInfo.Link.Type(),
		"link-name":      attrs.Name,
		"link-index":     attrs.Index,
		"link-mtu":       attrs.MTU,
		"link-hwaddr":    attrs.HardwareAddr,
		"link-encap":     attrs.EncapType,
		"link-operstate": attrs.OperState,
		"parent-dev":     attrs.ParentDev,
		"parent-dev-bus": attrs.ParentDevBus,
		"parent-index":   attrs.ParentIndex,
		"master-index":   attrs.MasterIndex,
		"alias":          attrs.Alias,
		"num-addrs":      len(netInfo.Addrs),
		"num-routes":     len(netInfo.Routes),
	}).Info("createPhysicalEndpoint: netInfo details for device correlation")

	sysIfaceDevicePath, bdf, err := getIfaceDevicePath(netInfo.Link, netInfo.Iface.Name)
	if err != nil {
		return nil, err
	}

	// For VMBus devices, attempt to discover a subordinate PCI VF
	// (e.g. mlx5 VF under netvsc with Azure Accelerated Networking).
	// When found, use VFIO passthrough on the VF instead of virtio-net.
	// When no subordinate VF is present, fall through to the virtio-net
	// network pair path.
	busType := netInfo.Link.Attrs().ParentDevBus
	isVFIO := (busType == "pci")
	if busType == "vmbus" {
		vfBDF, vfErr := findSubordinateVF(netInfo.Iface.Name)
		if vfErr != nil {
			networkLogger().WithError(vfErr).Warn("Error searching for subordinate VF, falling back to virtio-net pair")
		} else if vfBDF != "" {
			networkLogger().WithFields(logrus.Fields{
				"vmbus-guid": bdf,
				"vf-bdf":     vfBDF,
				"iface":      netInfo.Iface.Name,
			}).Info("Using subordinate PCI VF for VFIO passthrough instead of virtio-net pair")
			sysIfaceDevicePath = filepath.Join(sysBusPath, "pci", "devices", vfBDF)
			bdf = vfBDF
			busType = "pci"
			isVFIO = true
		}
	}

	// Get driver by following symlink /sys/bus/pci/devices/$bdf/driver or /sys/bus/vmbus/devices/$guid/driver
	driverPath := filepath.Join(sysIfaceDevicePath, "driver")
	link, err := os.Readlink(driverPath)
	if err != nil {
		return nil, err
	}
	driver := filepath.Base(link)
	// Get device by following symlink /sys/bus/pci/devices/$bdf/device or /sys/bus/vmbus/devices/$guid/device
	ifaceDevicePath := filepath.Join(sysIfaceDevicePath, "device")
	contents, err := os.ReadFile(ifaceDevicePath)
	if err != nil {
		return nil, err
	}
	deviceID := strings.TrimSpace(string(contents))
	// Vendor id (/sys/bus/pci/devices/$bdf/device or /sys/bus/vmbus/devices/$guid/driver)
	ifaceVendorPath := filepath.Join(sysIfaceDevicePath, "vendor")
	contents, err = os.ReadFile(ifaceVendorPath)
	if err != nil {
		return nil, err
	}
	netPair := NetworkInterfacePair{}
	if isVFIO {
		if isFVIODisabled {
			// When `cold_plug_vfio` is set to "no-port", the PhysicalEndpoint's VFIO device cannot be attached to the guest VM.
			// Fail early to prevent the interface from being unbound and rebound to the VFIO driver.
			return nil, fmt.Errorf("Unable to add PhysicalEndpoint %s because cold_plug_vfio is disabled", netInfo.Iface.Name)
		}
	} else {
		if idx < 0 {
			return nil, fmt.Errorf("invalid network endpoint index: %d", idx)
		}
		netPair, err = createNetworkInterfacePair(idx, netInfo.Iface.Name, interworkingModel)
		if err != nil {
			return nil, err
		}
		if netInfo.Iface.Name != "" {
			netPair.VirtIface.Name = netInfo.Iface.Name
		}
	}
	vendorID := strings.TrimSpace(string(contents))
	vendorDeviceID := fmt.Sprintf("%s %s", vendorID, deviceID)
	vendorDeviceID = strings.TrimSpace(vendorDeviceID)
	physicalEndpoint := &PhysicalEndpoint{
		IfaceName:      netInfo.Iface.Name,
		IsVFIO:         isVFIO,
		HardAddr:       netInfo.Iface.HardwareAddr.String(),
		VendorDeviceID: vendorDeviceID,
		EndpointType:   PhysicalEndpointType,
		Driver:         driver,
		BDF:            bdf,
		NetPair:        netPair,
		BusType:        busType,
	}
	return physicalEndpoint, nil
}
func bindNICToVFIO(endpoint *PhysicalEndpoint) (string, error) {
	return drivers.BindDevicetoVFIO(endpoint.BDF, endpoint.Driver)
}
func bindNICToHost(endpoint *PhysicalEndpoint) error {
	return drivers.BindDevicetoHost(endpoint.BDF, endpoint.Driver)
}
func (endpoint *PhysicalEndpoint) save() persistapi.NetworkEndpoint {
	netpair := saveNetIfPair(&endpoint.NetPair)
	return persistapi.NetworkEndpoint{
		Type: string(endpoint.Type()),
		Physical: &persistapi.PhysicalEndpoint{
			BDF:            endpoint.BDF,
			Driver:         endpoint.Driver,
			VendorDeviceID: endpoint.VendorDeviceID,
			NetPair:        *netpair,
			BusType:        endpoint.BusType,
			IsVFIO:         endpoint.IsVFIO,
		},
	}
}
func (endpoint *PhysicalEndpoint) load(s persistapi.NetworkEndpoint) {
	endpoint.EndpointType = PhysicalEndpointType
	if s.Physical != nil {
		netpair := loadNetIfPair(&s.Physical.NetPair)
		endpoint.NetPair = *netpair
		endpoint.BDF = s.Physical.BDF
		endpoint.Driver = s.Physical.Driver
		endpoint.VendorDeviceID = s.Physical.VendorDeviceID
		endpoint.BusType = s.Physical.BusType
		endpoint.IsVFIO = s.Physical.IsVFIO
	}
}

func (endpoint *PhysicalEndpoint) GetRxRateLimiter() bool {
	return endpoint.RxRateLimiter
}
func (endpoint *PhysicalEndpoint) SetRxRateLimiter() error {
	endpoint.RxRateLimiter = true
	return nil
}

func (endpoint *PhysicalEndpoint) GetTxRateLimiter() bool {
	return endpoint.TxRateLimiter
}
func (endpoint *PhysicalEndpoint) SetTxRateLimiter() error {
	endpoint.TxRateLimiter = true
	return nil
}

File: src/runtime-rs/crates/resource/src/network/endpoint/physical_endpoint.rs
// Copyright (c) 2019-2022 Alibaba Cloud
// Copyright (c) 2019-2022 Ant Group
//
// SPDX-License-Identifier: Apache-2.0
//

use std::io::{self, Error};
use std::path::Path;
use std::sync::Arc;

use anyhow::{anyhow, Context, Result};
use async_trait::async_trait;
use hypervisor::device::device_manager::{do_handle_device, DeviceManager};
use hypervisor::device::driver::NetworkConfig;
use hypervisor::device::{DeviceConfig, DeviceType};
use hypervisor::{device::driver, Hypervisor, NetworkDevice};
use hypervisor::{get_vfio_device, VfioConfig};
use tokio::sync::RwLock;

use super::endpoint_persist::{EndpointState, PhysicalEndpointState};
use super::Endpoint;
use crate::network::network_pair::{NetworkInterface, NetworkPair};
use crate::network::utils::{self, link};

#[derive(Debug)]
pub struct VendorDevice {
    vendor_id: String,
    device_id: String,
}

impl VendorDevice {
    pub fn new(vendor_id: &str, device_id: &str) -> Result<Self> {
        if vendor_id.is_empty() || device_id.is_empty() {
            return Err(anyhow!(
                "invalid parameters vendor_id {} device_id {}",
                vendor_id,
                device_id
            ));
        }
        Ok(Self {
            vendor_id: vendor_id.to_string(),
            device_id: device_id.to_string(),
        })
    }

    pub fn vendor_device_id(&self) -> String {
        format!("{}_{}", &self.vendor_id, &self.device_id)
    }
}

#[derive(Debug)]
pub struct PhysicalEndpoint {
    iface_name: String,
    hard_addr: String,
    bdf: String,
    driver: String,
    vendor_device_id: VendorDevice,
    is_vf: bool,
    bus_type: link::BusType,
    net_pair: NetworkPair,
    d: Arc<RwLock<DeviceManager>>,
}

impl PhysicalEndpoint {
    /// Create a new PhysicalEndpoint.
    ///
    /// For VF (SR-IOV Virtual Function) devices, the NIC will be passed through
    /// to the VM via VFIO — no tap/bridge pair is needed, so we create a minimal
    /// stub NetworkPair.
    ///
    /// For non-VF physical NICs, we need a real tap+bridge pair (like veth
    /// endpoints use) so that traffic can flow between the host NIC and the VM
    /// through the hypervisor's TAP backend. The `handle`, `idx`, `model`, and
    /// `queues` parameters are required to create this pair.
    pub async fn new(
        handle: &rtnetlink::Handle,
        name: &str,
        hardware_addr: &[u8],
        idx: u32,
        model: &str,
        queues: usize,
        d: Arc<RwLock<DeviceManager>>,
    ) -> Result<Self> {
        // Determine bus type (PCI or VMBus) and resolve the sysfs device path.
        // For PCI: uses ethtool to get BDF, path = /sys/bus/pci/devices/<bdf>
        // For VMBus: resolves device symlink, path = /sys/bus/vmbus/devices/<guid>
        let (sys_iface_device_path, bdf, bus_type) =
            link::get_iface_device_path(name).context("get iface device path")?;
        let sys_device_path = Path::new(&sys_iface_device_path);

        // Get driver by following symlink <device_path>/driver
        let driver_path = sys_device_path.join("driver");
        let link_target = driver_path.read_link().context("read link")?;
        let driver = link_target
            .file_name()
            .map_or(String::new(), |v| v.to_str().unwrap().to_owned());

        // Get vendor and device id from sysfs device path
        let iface_device_path = sys_device_path.join("device");
        let device_id = std::fs::read_to_string(&iface_device_path)
            .with_context(|| format!("read device path {:?}", &iface_device_path))?;

        let iface_vendor_path = sys_device_path.join("vendor");
        let vendor_id = std::fs::read_to_string(&iface_vendor_path)
            .with_context(|| format!("read vendor path {:?}", &iface_vendor_path))?;
        let is_vf = link::is_vf(name).context("check if is vf")?;

        // VF devices use VFIO passthrough — no real tap/bridge needed.
        // Non-VF devices need a real tap+bridge pair for the hypervisor to
        // connect the host NIC traffic into the VM, matching Go's
        // createNetworkInterfacePair() call in createPhysicalEndpoint().
        let net_pair = if is_vf {
            NetworkPair::new_for_physical(name, hardware_addr, is_vf)
                .context("new network pair for physical vf endpoint")?
        } else {
            NetworkPair::new(handle, idx, name, model, queues)
                .await
                .context("new network pair for physical non-vf endpoint")?
        };

        Ok(Self {
            iface_name: name.to_string(),
            hard_addr: utils::get_mac_addr(hardware_addr).context("get mac addr")?,
            vendor_device_id: VendorDevice::new(&vendor_id, &device_id)
                .context("new vendor device")?,
            driver,
            bdf,
            is_vf,
            bus_type,
            net_pair,
            d,
        })
    }

    pub fn network_pair(&self) -> &NetworkPair {
        &self.net_pair
    }

    fn get_network_config(&self) -> Result<NetworkConfig> {
        let iface = &self.net_pair.tap.tap_iface;
        let guest_mac = utils::parse_mac(&iface.hard_addr).ok_or_else(|| {
            Error::new(
                io::ErrorKind::InvalidData,
                format!("hard_addr {}", &iface.hard_addr),
            )
        })?;

        Ok(NetworkConfig {
            host_dev_name: iface.name.clone(),
            virt_iface_name: self.net_pair.virt_iface.name.clone(),
            guest_mac: Some(guest_mac),
            ..Default::default()
        })
    }
}

#[async_trait]
impl Endpoint for PhysicalEndpoint {
    async fn name(&self) -> String {
        self.iface_name.clone()
    }

    async fn hardware_addr(&self) -> String {
        self.hard_addr.clone()
    }

    async fn attach(&self) -> Result<()> {
        if self.is_vf {
            // bind physical interface from host driver and bind to vfio
            driver::bind_device_to_vfio(
                &self.bdf,
                &self.driver,
                &self.vendor_device_id.vendor_device_id(),
            )
            .with_context(|| format!("bind physical endpoint from {} to vfio", &self.driver))?;

            let vfio_device =
                get_vfio_device(self.bdf.clone()).context("get vfio device failed.")?;
            let vfio_dev_config = &mut VfioConfig {
                host_path: vfio_device.clone(),
                dev_type: "pci".to_string(),
                hostdev_prefix: "physical_nic_".to_owned(),
                ..Default::default()
            };

            // create and insert VFIO device into Kata VM
            do_handle_device(&self.d, &DeviceConfig::VfioCfg(vfio_dev_config.clone()))
                .await
                .context("do handle device failed.")?;

            Ok(())
        } else {
            self.net_pair
                .add_network_model()
                .await
                .context("add network model")?;
            let config = self.get_network_config().context("get network config")?;
            do_handle_device(&self.d, &DeviceConfig::NetworkCfg(config))
                .await
                .context("do handle network Physical endpoint device failed.")?;
            Ok(())
        }
    }

    // detach for physical endpoint unbinds the physical network interface from vfio-pci
    // and binds it back to the saved host driver.
    async fn detach(&self, _hypervisor: &dyn Hypervisor) -> Result<()> {
        if self.is_vf {
            driver::bind_device_to_host(
                &self.bdf,
                &self.driver,
                &self.vendor_device_id.vendor_device_id(),
            )
            .with_context(|| {
                format!(
                    "bind physical endpoint device from vfio to {}",
                    &self.driver
                )
            })?;
            Ok(())
        } else {
            self.net_pair
                .del_network_model()
                .await
                .context("del network model")?;
            let config = self.get_network_config().context("get network config")?;
            _hypervisor
                .remove_device(DeviceType::Network(NetworkDevice {
                    config,
                    ..Default::default()
                }))
                .await
                .context("remove Physical endpoint device by hypervisor failed.")?;
            Ok(())
        }
    }

    async fn save(&self) -> Option<EndpointState> {
        Some(EndpointState {
            physical_endpoint: Some(PhysicalEndpointState {
                bdf: self.bdf.clone(),
                driver: self.driver.clone(),
                vendor_id: self.vendor_device_id.vendor_id.clone(),
                device_id: self.vendor_device_id.device_id.clone(),
                hard_addr: self.hard_addr.clone(),
                is_vf: self.is_vf,
                iface_name: self.iface_name.clone(),
                bus_type: match self.bus_type {
                    link::BusType::Pci => "pci".to_string(),
                    link::BusType::Vmbus => "vmbus".to_string(),
                },
            }),
            ..Default::default()
        })
    }
}
